homepage/content/blog/2021-03-05-switchMap-concatMap-mergemap-exhaustmap-explained.md

```diff
@@ -0,0 +1,264 @@
+
+  // After subscribing we emit the two value in the observable, could also be more than that
+  this.sub.next("first");
+  this.sub.next("second");
+
+  console.log("-------");
+}
+```
+
+The `switchMap` operator takes the first value in the stream `first` and calls the `anyLongRunningOp` with it. _Right after_ it did this it receives the second emit with the value `second`. Now it forgets about the response of the first request. It is not waiting for it. Like the HTTP request which could be done here is out and comes back any when but the `switchMap` operator does not care about the first one. It calls the `anyLongRunningOp` with the `second` parameter and waits for _that one_'s answer. And so with multiple ones, it is only interested in the response of the _last_ one it fired. Everything before got ignored.
+
+![SwitchMap operator](https://cdn.offering.solutions/img/articles/2021-03-07/switchmap.gif)
+
+In the animation you can see that only the value `second` is printed. This is the last one which got fired. The first response is ignored by the `switchMap` operator when a second one comes in and the first one is not finished yet.
+
+## ConcatMap
+
+Next one in the list is the `concatMap` operator. Let us assume the same method again:
+
+```ts
+fireEvents() {
+
+  // Here we react ot everything which is fired in the subject
+  this.sub
+    // Here we can take the operator we want to take a look at which returns the
+    // result from the anyLongRunningOp method which is the value itself
+    // (for the sake of simplicity). The `tap` only prints out what
+    // was just sent, does nothing to the stream!
+    .pipe(
+      tap(value => console.log("--> sent out", value)),
+      concatMap(value => this.obsService.anyLongRunningOp(value))
+    )
+    // We console.log the output, which is 'first' or 'second' or
+    .subscribe(value => console.log("<-- received", value));
+
+  // After subscribing we emit the two value in the observable, could also be more than that
+  this.sub.next("first");
+  this.sub.next("second");
+
+  console.log("-------");
+}
+```
+
+We know that the `switchMap` operator is only interested in the most recent value which came in. It does not build a relation between everything which comes in and puts them in a queue. This is what the `concatMap` operator is for. It behaves like it has a queue and stores the incoming calls and emits the next one when the previous one came back! So when it receives the value with `first` it calls `anyLongRunningOp` with `first`, then the `second` value comes in. The `concatMap` operator now holds this call back until the `anyLongRunningOp` method comes back with the result of the call with `first` and _then_ the next call with `second` as parameter is being fired. It concatenates the calls and emits them one after another. As a side effect it builds a relation between the calls because it has to look wether the first one came back before it can emit the next one.

```

Comment:
has a queue and stores the incoming calls

FYI: Das ist gleichzeitig auch das größte Problem mit `concatMap`, da man eine BufferOverflow exception bekommen kann, wenn man zu viele Werte emitted, ohne dass die inneren observables completen